frappe/deprecation_dumpster.py

```diff
@@ -0,0 +1,288 @@
+"""
+Welcome to the Deprecation Dumpster: Where Old Code Goes to Party! 🎉🗑️
+
+This file is the final resting place (or should we say, "retirement home"?) for all the deprecated functions and methods of the Frappe framework. It's like a code nursing home, but with more monkey-patching and less bingo.
+
+Each function or method that checks in here comes with its own personalized decorator, complete with:
+1. The date it was marked for deprecation (its "over the hill" birthday)
+2. The Frappe version in which it will be removed (its "graduation" to the great codebase in the sky)
+3. A user-facing note on alternative solutions (its "parting wisdom")
+
+Warning: The global namespace herein is more patched up than a sailor's favorite pair of jeans. Proceed with caution and a sense of humor!
+
+Remember, deprecated doesn't mean useless - it just means these functions are enjoying their golden years before their final bow. Treat them with respect, and maybe bring them some virtual prune juice.
+
+Enjoy your stay in the Deprecation Dumpster, where every function gets a second chance to shine (or at least, to not break everything).
+"""
+
+import inspect
+import os
+import sys
+import warnings
+
+
+def is_tty():
+	return sys.stdout.isatty()
+
+
+def colorize(text, color_code):
+	if is_tty():
+		return f"\033[{color_code}m{text}\033[0m"
+	return text
+
+
+try:
+	# since python 3.13, PEP 702
+	from warnings import deprecated as _deprecated
+except Exception:
+	import functools
+	import warnings
+	from collections.abc import Callable
+	from typing import Optional, TypeVar, Union, overload
+
+	T = TypeVar("T", bound=Callable)
+
+	def _deprecated(message: str, category=DeprecationWarning, stacklevel=1) -> Callable[[T], T]:
+		def decorator(func: T) -> T:
+			@functools.wraps(func)
+			def wrapper(*args, **kwargs):
+				if message:
+					warning_msg = f"{func.__name__} is deprecated.\n{message}"
+				else:
+					warning_msg = f"{func.__name__} is deprecated."
+				warnings.warn(warning_msg, category=category, stacklevel=stacklevel + 1)
+				return func(*args, **kwargs)
+
+			return wrapper
+			wrapper.__deprecated__ = True  # hint for the type checker
+
+		return decorator
+
+
+def deprecated(original: str, marked: str, graduation: str, msg: str, stacklevel: int = 1):
+	"""Decorator to wrap a function/method as deprecated.
+
+	Arguments:
+	        - original: frappe.utils.make_esc  (fully qualified)
+	        - marked: 2024-09-13  (the date it has been marked)
+	        - graduation: v17  (generally: current version + 2)
+	"""
+
+	def decorator(func):
+		# Get the filename of the caller
+		frame = inspect.currentframe()
+		caller_filepath = frame.f_back.f_code.co_filename
+		if os.path.basename(caller_filepath) != "deprecation_dumpster.py":
+			raise RuntimeError(
+				colorize("The deprecated function ", "93")
```

Comment:
Could you extract the colour codes into a separate class and reference that here, could be easier to understand.

maybe something like

```python
class Color:
    CYAN = 96
```

Reply:
https://github.com/frappe/frappe/pull/27770/commits/675e9b4a00ac60b1740cb77e203b2e56b6841513